File: src/eko-core/prompt/components/registry.ts
/**
 * 组件注册表
 * 管理所有提示词组件的注册和获取
 */

import { ComponentRegistry, ComponentConfig, PromptComponent } from './types'
import * as agentComponents from './agent'
import * as systemComponents from './system'
import * as toolComponents from './tools'
import * as taskComponents from './task'
import * as planningComponents from './planning'
import * as dialogueComponents from './dialogue'
import * as workspaceComponents from './workspace'

/**
 * 全局组件注册表
 */
class PromptComponentRegistry {
  private static instance: PromptComponentRegistry
  private registry: ComponentRegistry = {} as ComponentRegistry
  private loaded = false

  static getInstance(): PromptComponentRegistry {
    if (!PromptComponentRegistry.instance) {
      PromptComponentRegistry.instance = new PromptComponentRegistry()
    }
    return PromptComponentRegistry.instance
  }

  /**
   * 加载所有组件
   */
  async load(): Promise<void> {
    if (this.loaded) return

    // 注册各类组件
    const isComponentConfig = (v: unknown): v is ComponentConfig => {
      return (
        typeof v === 'object' &&
        v !== null &&
        'id' in (v as Record<string, unknown>) &&
        'fn' in (v as Record<string, unknown>)
      )
    }

    const all = [
      ...Object.values(agentComponents),
      ...Object.values(systemComponents),
      ...Object.values(toolComponents),
      ...Object.values(taskComponents),
      ...Object.values(planningComponents),
      ...Object.values(dialogueComponents),
      ...Object.values(workspaceComponents),
    ]

    this.registerComponents(all.filter(isComponentConfig))

    this.loaded = true
  }

  /**
   * 注册组件
   */
  private registerComponents(components: ComponentConfig[]): void {
    for (const component of components) {
      this.registry[component.id] = component
    }
  }

  /**
   * 获取组件
   */
  getComponent(id: PromptComponent): ComponentConfig | undefined {
    return this.registry[id]
  }

  /**
   * 获取所有组件
   */
  getAllComponents(): ComponentRegistry {
    return { ...this.registry }
  }

  /**
   * 检查组件是否存在
   */
  hasComponent(id: PromptComponent): boolean {
    return id in this.registry
  }

  /**
   * 获取组件列表
   */
  getComponentIds(): PromptComponent[] {
    return Object.keys(this.registry) as PromptComponent[]
  }

  /**
   * 根据依赖关系排序组件
   */
  sortComponentsByDependencies(components: PromptComponent[]): PromptComponent[] {
    const sorted: PromptComponent[] = []
    const visited = new Set<PromptComponent>()
    const visiting = new Set<PromptComponent>()

    const visit = (componentId: PromptComponent) => {
      if (visiting.has(componentId)) {
        throw new Error(`检测到循环依赖: ${componentId}`)
      }
      if (visited.has(componentId)) {
        return
      }

      visiting.add(componentId)
      const component = this.registry[componentId]

      if (component?.dependencies) {
        for (const dep of component.dependencies) {
          if (components.includes(dep)) {
            visit(dep)
          }
        }
      }

      visiting.delete(componentId)
      visited.add(componentId)
      sorted.push(componentId)
    }

    for (const componentId of components) {
      if (!visited.has(componentId)) {
        visit(componentId)
      }
    }

    return sorted
  }

  /**
   * 验证组件依赖
   */
  validateDependencies(components: PromptComponent[]): string[] {
    const errors: string[] = []

    for (const componentId of components) {
      const component = this.registry[componentId]
      if (!component) {
        errors.push(`组件不存在: ${componentId}`)
        continue
      }

      if (component.dependencies) {
        for (const dep of component.dependencies) {
          if (!components.includes(dep)) {
            errors.push(`组件 ${componentId} 依赖的组件 ${dep} 未包含在列表中`)
          }
        }
      }
    }

    return errors
  }
}

/**
 * 获取组件注册表实例
 */
export function getComponentRegistry(): PromptComponentRegistry {
  return PromptComponentRegistry.getInstance()
}

/**
 * 便捷函数：获取组件
 */
export async function getComponent(id: PromptComponent): Promise<ComponentConfig | undefined> {
  const registry = getComponentRegistry()
  await registry.load()
  return registry.getComponent(id)
}

/**
 * 便捷函数：获取所有组件
 */
export async function getAllComponents(): Promise<ComponentRegistry> {
  const registry = getComponentRegistry()
  await registry.load()
  return registry.getAllComponents()
}

File: src-tauri/src/agent/prompt/components/registry.rs
use std::collections::HashSet;
use std::sync::Arc;

use crate::agent::config::PromptComponent;
use crate::agent::prompt::components::types::{ComponentDefinition, ComponentRegistry};

use super::{agent, dialogue, planning, system, task, tools, workspace};

/// Runtime registry mirroring the front-end component registry.
pub struct PromptComponentRegistry {
    components: ComponentRegistry,
    loaded: bool,
}

impl PromptComponentRegistry {
    pub fn new() -> Self {
        Self {
            components: ComponentRegistry::new(),
            loaded: false,
        }
    }

    /// Lazily load all component families.
    pub fn ensure_loaded(&mut self) {
        if self.loaded {
            return;
        }

        self.register_many(agent::definitions());
        self.register_many(system::definitions());
        self.register_many(tools::definitions());
        self.register_many(task::definitions());
        self.register_many(planning::definitions());
        self.register_many(dialogue::definitions());
        self.register_many(workspace::definitions());

        self.loaded = true;
    }

    fn register_many(&mut self, defs: Vec<Arc<dyn ComponentDefinition>>) {
        for def in defs {
            self.components.insert(def.id(), Arc::clone(&def));
        }
    }

    pub fn get(&mut self, id: PromptComponent) -> Option<Arc<dyn ComponentDefinition>> {
        self.ensure_loaded();
        self.components.get(&id).map(Arc::clone)
    }

    pub fn get_all(&mut self) -> ComponentRegistry {
        self.ensure_loaded();
        self.components.clone()
    }

    pub fn sort_by_dependencies(
        &mut self,
        components: &[PromptComponent],
    ) -> Result<Vec<PromptComponent>, String> {
        self.ensure_loaded();

        let mut sorted = Vec::new();
        let mut visited = HashSet::new();
        let mut visiting = HashSet::new();

        fn visit(
            id: PromptComponent,
            registry: &ComponentRegistry,
            components: &[PromptComponent],
            sorted: &mut Vec<PromptComponent>,
            visited: &mut HashSet<PromptComponent>,
            visiting: &mut HashSet<PromptComponent>,
        ) -> Result<(), String> {
            if visiting.contains(&id) {
                return Err(format!("Circular dependency detected: {:?}", id));
            }
            if visited.contains(&id) {
                return Ok(());
            }

            visiting.insert(id.clone());
            if let Some(def) = registry.get(&id) {
                for dep in def.dependencies() {
                    if components.contains(dep) {
                        visit(dep.clone(), registry, components, sorted, visited, visiting)?;
                    }
                }
            }

            visiting.remove(&id);
            visited.insert(id.clone());
            sorted.push(id.clone());
            Ok(())
        }

        for component in components {
            if !visited.contains(component) {
                visit(
                    component.clone(),
                    &self.components,
                    components,
                    &mut sorted,
                    &mut visited,
                    &mut visiting,
                )?;
            }
        }

        Ok(sorted)
    }

    pub fn validate_dependencies(&mut self, components: &[PromptComponent]) -> Vec<String> {
        self.ensure_loaded();

        let mut errors = Vec::new();
        for component in components {
            match self.components.get(component) {
                Some(def) => {
                    for dep in def.dependencies() {
                        if !components.contains(dep) {
                            errors.push(format!(
                                "Component {:?} depends on {:?} which is missing from the selection",
                                component, dep
                            ));
                        }
                    }
                }
                None => errors.push(format!("Component does not exist: {:?}", component)),
            }
        }
        errors
    }
}

impl Default for PromptComponentRegistry {
    fn default() -> Self {
        Self::new()
    }
}
